src/url.rs

```diff
@@ -0,0 +1,387 @@
+
+use oxigraph::{
+    model::{GraphName, NamedNodeRef, NamedOrBlankNode, Quad, Term},
+    store::Store,
+};
+use url::Url;
+
+use reqwest::blocking::Client;
+
+use cached::proc_macro::cached;
+use cached::Return;
+
+use crate::{
+    rdf::{
+        add_url_status_metric, create_metrics_store, dump_graph_as_turtle,
+        extract_urls_from_distribution, get_dataset_node, list_distributions, parse_turtle,
+    },
+    schemas::{MQAEvent, MQAEventType},
+};
+
+use crate::error::Error;
+
+#[derive(Debug, Clone)]
+pub enum UrlType {
+    AccessUrl,
+    DownloadUrl,
+}
+
+#[derive(Debug, Clone)]
+pub struct UrlCheck {
+    pub method: String,
+    pub url_type: UrlType,
+    pub url: String,
+}
+#[derive(Debug, Clone)]
+pub struct UrlCheckResult {
+    pub url: String,
+    pub url_type: UrlType,
+    pub status: u16,
+    pub note: String,
+}
+
+pub fn parse_rdf_graph_and_check_urls(
+    fdk_id: String,
+    graph: String,
+) -> Result<Option<MQAEvent>, String> {
+    match parse_turtle(graph) {
+        Ok(store) => {
+            match get_dataset_node(&store) {
+                Some(dataset_node) => {
+                    match check_urls(fdk_id.clone(), dataset_node.as_ref(), &store) {
+                        Ok(metrics_store) => {
+                            match dump_graph_as_turtle(&metrics_store) {
+                                Ok(bytes) => {
+                                    // Create MQA event
+                                    match std::str::from_utf8(bytes.as_slice()) {
+                                        Ok(turtle) => Ok(Some(MQAEvent {
+                                            event_type: MQAEventType::UrlsChecked,
+                                            fdk_id: fdk_id.clone(),
+                                            graph: turtle.to_string(),
+                                            timestamp: Utc::now().timestamp_millis(),
+                                        })),
+                                        Err(e) => Err(format!(
+                                            "{} - Failed dumping graph as turtle: {}",
+                                            fdk_id, e
+                                        )),
+                                    }
+                                }
+                                Err(e) => Err(format!(
+                                    "{} - Failed dumping graph as turtle: {}",
+                                    fdk_id, e
+                                )),
+                            }
+                        }
+                        Err(e) => Err(format!(
+                            "{} - Failed creating new metrics store: {}",
+                            fdk_id, e
+                        )),
+                    }
+                }
+                None => Err(format!("{} - Dataset node not found in graph", fdk_id)),
+            }
+        }
+        Err(e) => Err(format!("{} - Failed to parse graph: {}", fdk_id, e)),
+    }
```

Comment:
Synes det er mye mer lesbart om man bruker `?`. Trenger nok også ikke å konvertere feil til strings der feilen i seg selv sier tilstrekkelig om problemet. Å legge på `fdk_id` ville jeg gjort på stedet som kaller funksjonen, så man slipper å gjøre det på alle feilene: `parse_rdf_graph_and_check_urls(fdk_id, graph).map_err(|e| format!("{} - {}", fdk_id, e))`

```suggestion
) -> Result<Option<MQAEvent>, Error> {
    let store = parse_turtle(graph)?;
    let dataset_node = get_dataset_node(&store).ok_or("Dataset node not found in graph")?;
    let metrics_store = check_urls(fdk_id.clone(), dataset_node.as_ref(), &store)?;
    let bytes = dump_graph_as_turtle(&metrics_store)?;
    let turtle = std::str::from_utf8(bytes.as_slice())
        .map_err(|e| format!("Failed converting graph to string: {}", e))?;

    Ok(Some(MQAEvent {
        event_type: MQAEventType::UrlsChecked,
        fdk_id: fdk_id.clone(),
        graph: turtle.to_string(),
        timestamp: Utc::now().timestamp_millis(),
    }))
```